data-pipeline-flink/notification-engine/src/main/scala/org/sunbird/dp/notification/function/LatestCourseEmailNotificationFunction.scala

```diff
@@ -0,0 +1,339 @@
+        val excludeFields=new Array[String](0)
+        sourceBuilder.fetchSource(courseConfig.fields.split(",",-1),excludeFields)
+        sourceBuilder.from(offset)
+        sourceBuilder.size(45)
+        val index = new IndexService(courseConfig)
+        response = index.getEsResult(courseConfig.sb_es_user_profile_index, courseConfig.es_profile_index_type, sourceBuilder, true)
+        response.getHits.forEach(hitDetails => {
+          result = hitDetails.getSourceAsMap().asInstanceOf[util.HashMap[String, Any]]
+          val userId = result.get(courseConfig.USERID).toString
+          if (result.containsKey(courseConfig.PROFILE_DETAILS)) {
+            val profileDetails: util.HashMap[String, Any] = result.get(courseConfig.PROFILE_DETAILS).asInstanceOf[util.HashMap[String, Any]]
+            if (profileDetails.containsKey(courseConfig.PERSONAL_DETAILS)) {
+              val personalDetails: util.HashMap[String, Any] = profileDetails.get(courseConfig.PERSONAL_DETAILS).asInstanceOf[util.HashMap[String, Any]]
+              if (MapUtils.isNotEmpty(personalDetails)) {
+                val email: String = personalDetails.get(courseConfig.PRIMARY_EMAIL).asInstanceOf[String]
+                if (StringUtils.isNotBlank(email) && !excludeEmailsList.contains(email)) {
+                  if (courseConfig.MAIL_LIST != null && !courseConfig.MAIL_LIST.contains(email)) {
+                    val emailWithUserIdMap = new util.HashMap[String,Any]()
+                    emailWithUserIdMap.put(userId,email)
+                    emailWithUserId.add(emailWithUserIdMap)
+                  } else {
+                    logger.info("Invalid Email :" + email)
+                  }
+                }
+              }
+            }
+          }
+        })
+        if (CollectionUtils.isNotEmpty(emailWithUserId)) {
+          val kafkaProducerProps = new Properties()
+          kafkaProducerProps.put(courseConfig.bootstrap_servers, courseConfig.BOOTSTRAP_SERVER_CONFIG)
+          kafkaProducerProps.put(courseConfig.key_serializer, classOf[StringSerializer])
+          kafkaProducerProps.put(courseConfig.value_serializer, classOf[StringSerializer])
+          val producer = new KafkaProducer[String, String](kafkaProducerProps)
+          val producerData = new util.HashMap[String, Any]
+          producerData.put(courseConfig.MESSAGE,courseConfig.CHECK_NOTIFICATION_PREFERENCE_KEY)
+          producerData.put(courseConfig.emailWithUserId,emailWithUserId)
+          producerData.put(courseConfig.PARAMS,params)
+          producerData.put(courseConfig.emailTemplate, courseConfig.NEW_COURSES)
+          producerData.put(courseConfig.emailSubject, courseConfig.NEW_COURSES_MAIL_SUBJECT)
+          val gson=new Gson()
+          val jsonString=gson.toJson(producerData)
+          producer.send(new ProducerRecord[String, String](courseConfig.notification_Preference_Kafka_input_topic,  courseConfig.DATA, jsonString))
```

Comment:
yes there is no issue actually for incomplete course and first course we directly send the message to kafka to send email but for latest course we have a filter flink job also there so it will first filtered the user who disabled notification and then send to the kafka topic

Reply:
ok, thanks for clarifying